auth/token_verifier.go

```diff
@@ -34,8 +34,213 @@ import (
+	projectID         string
+	issuerPrefix      string
+	keySource         keySource
+	clock             internal.Clock
+}
+
+func newIDTokenVerifier(ctx context.Context, projectID string) (*tokenVerifier, error) {
+	noAuthHTTPClient, _, err := transport.NewHTTPClient(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return nil, err
+	}
+	return &tokenVerifier{
+		shortName:         "ID token",
+		articledShortName: "an ID token",
+		docURL:            "https://firebase.google.com/docs/auth/admin/verify-id-tokens",
+		projectID:         projectID,
+		issuerPrefix:      idTokenIssuerPrefix,
+		keySource:         newHTTPKeySource(idTokenCertURL, noAuthHTTPClient),
+		clock:             internal.SystemClock,
+	}, nil
+}
+
+// VerifyToken Verifies that the given token string is a valid Firebase JWT.
+//
+// VerifyToken considers a token string to be valid if all the following conditions are met:
+//   - The token string is a valid RS256 JWT.
+//   - The JWT contains a valid key ID (kid) claim.
+//   - The JWT contains valid issuer (iss) and audience (aud) claims that match the issuerPrefix
+//     and projectID of the tokenVerifier.
+//   - The JWT contains a valid subject (sub) claim.
+//   - The JWT is not expired, and it has been issued some time in the past.
+//   - The JWT is signed by a Firebase Auth backend server as determined by the keySource.
+//
+// If any of the above conditions are not met, an error is returned. Otherwise a pointer to a
+// decoded Token is returned.
+func (tv *tokenVerifier) VerifyToken(ctx context.Context, token string) (*Token, error) {
+	if tv.projectID == "" {
+		return nil, errors.New("project id not available")
+	}
+	if token == "" {
+		return nil, fmt.Errorf("%s must be a non-empty string", tv.shortName)
+	}
+
+	// Validate the token content first. This is fast and cheap.
+	payload, err := tv.verifyContent(token)
+	if err != nil {
+		return nil, fmt.Errorf("%s; see %s for details on how to retrieve a valid %s",
+			err.Error(), tv.docURL, tv.shortName)
+	}
+
+	if err := tv.verifyTimestamps(payload); err != nil {
+		return nil, err
+	}
+
+	// Verifying the signature requires syncronized access to a key cache and
+	// potentially issues an http request. Therefore we do it last.
+	if err := tv.verifySignature(ctx, token); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func (tv *tokenVerifier) verifyContent(token string) (*Token, error) {
+	var (
+		header  jwtHeader
+		payload Token
+		err     error
+	)
+
+	segments := strings.Split(token, ".")
+	if len(segments) != 3 {
+		return nil, errors.New("incorrect number of segments")
+	}
+
+	if err = decode(segments[0], &header); err != nil {
+		return nil, err
+	}
+
+	if err = decode(segments[1], &payload); err != nil {
+		return nil, err
+	}
+
+	issuer := tv.issuerPrefix + tv.projectID
+	if header.KeyID == "" {
+		if payload.Audience == firebaseAudience {
+			err = fmt.Errorf("expected %s but got a custom token", tv.articledShortName)
```

Comment:
I think all of this would be much more idiomatic if there were no `err` variable and it instead used `return` in these cases, and you got rid of all the `else`s. That'll also make it clearer / ensure that there's no logic happening in this function once the error condition has been determined.

https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

Reply:
Done